samples/excel/90-scenarios/currency-converter.yaml

```diff
@@ -0,0 +1,178 @@
+order: 5
+id: excel-scenarios-currency-converter
+name: Currency Converter
+description: 'Uses an exchange rate API to convert currency values based on their original transaction times.'
+author: cakriwut
+host: EXCEL
+api_set:
+    ExcelApi: '1.4'
+script:
+    content: |
+        /*
+        * Copyright (c) Riwut Libinuko. All rights reserved. Licensed under the MIT license.
+        */
+
+        declare let moment: any;
+        const tableName = "TransactionTable";
+
+        let tblSetup: Record<string, number> = {};
+
+        $("#convert").click(() => tryCatch(convert));
+        $("#setup").click(() => tryCatch(setupSample));
+
+        /** Main converting function **/
+        async function convert() {
+          await Excel.run(async (context) => {
+            const table = context.workbook.tables.getItem(tableName)
+            table.columns.load("items");
+            await context.sync();
+            if (!IsTableValid(table.columns.items)) {
+              console.error('Error: Some of default table header is missing. Required headers: \r\n"Currency","Price (Currency)","Transaction Date","Base","Price in Base"')
+              return;
+            }
+
+            table.rows.load("items,values");
+            await context.sync();
+            asyncForEach(table.rows.items, async (row,idx,arr) => {
+            const priceBaseRange = row.getRange().getCell(0, tblSetup['Price in Base']);
+
+            priceBaseRange.load("address,text");
+            context.trackedObjects.add(priceBaseRange);
+            await context.sync();
+            const priceInBase = row.values[0][tblSetup['Price in Base']];
+            const priceInCurrency = row.values[0][tblSetup['Price (Currency)']];
+            const transactionDate = row.values[0][tblSetup['Transaction Date']]
+            const currency = row.values[0][tblSetup['Currency']];
+            const baseCurrency = row.values[0][tblSetup['Base']]
+            
+            if (priceInBase === "") {
+              const dateMoment = moment.fromOADate(transactionDate);
+              const period = dateMoment.format("YYYY-MM-DD");
+              const queryUrl = convertValue(currency, baseCurrency, period, period);
+              const response = await fetch(queryUrl);
+              if (response.ok) {
+                const result = await response.json();
+                const converted = priceInCurrency * result.rates[baseCurrency];
+                priceBaseRange.values = [[converted]];
+                await context.sync();
+                context.trackedObjects.remove(priceBaseRange);
+              }
+            } else {
+                console.info(`Row: ${idx} has been converted. Skipped.`)
+              }
+            });
+          });
+        }
+
+        /** Async ForEeach */
+        async function asyncForEach(array: Array<any>, callback) {
+          for(let idx = 0; idx < array.length; idx++) {
+            await callback(array[idx],idx, array);
```

Comment:
The function name is misleading, since the for each isn't asynchronous. Please correct me if I'm wrong, but the `await` makes each callback complete before proceeding. I'm not sure you need all this either. Wasn't the original issue that we just needed to await a `context.sync()`?

Reply:
The original Array.ForEach(async () => {}) , does not await context.sync(). I tested using following test step.
1. Add 20 rows
2. Click "Convert"  ==> OK
3. Delete the converted value (Price In Base)
4. Click "Convert" ==> ERROR

The error - 
--
message: "The object path 'tables.getItem().rows.getItemAt().getRange().getCell()' isn't working for what you're trying to do. If you're using the object across multiple "context.sync" calls and outside the sequential execution of a ".run" batch, please use the "context.trackedObjects.add()" and "context.trackedObjects.remove()" methods to manage the object's lifetime."
---

Reply:
@cakriwut  Do you need to track the objects? All of this is in the same `Excel.run` call and uses the same context object. Check out this gist that removes the tracking and async for loop: https://gist.github.com/AlexJerabek/600b32fbc2787a13367d8b65ec133f16

I tested it with 120+ rows and never hit a race condition.

Reply:
@AlexJerabek , it is strange that you didn't face the issue, in my Excel (Microsoft Office Professional Plus 2019, version 1911 build 12228.20250) - I have the issue in the next run.
My steps
1. Create 50 samples row
2. Click Convert  (I found no issue in this step)
3. Clear "Price In Base"
4. Click Convert (the race condition appears).

See the video here,  - https://drive.google.com/file/d/1QDCNrS39XDiixrRegtAFgQhNKMCPKSvX/view?usp=sharing

Reply:
@AlexJerabek , I replace using for( let item of Array) instead. This solve the await inside Array.forEach